src/pass/verify_memory.cc

```diff
@@ -0,0 +1,166 @@
+  MemoryAccessVerifier &operator=(MemoryAccessVerifier &&) = delete;
+  //@}
+
+  /// Interface to perform memory access verification
+  void Run() {
+    if (kDLGPU != dev_type_) return;
+    IRVisitor::Visit(func_->body);
+  }
+
+  /// Verification result
+  inline bool Failed() const { return failure_; }
+
+ protected:
+  /// Visitor implementation
+  //@{
+  void Visit(const NodeRef &n) final {
+    if (Failed()) return;
+    IRVisitor::Visit(n);
+  }
+
+  void Visit_(const LetStmt *op) final {
+    // Book keep definitions
+    defs_[op->var.get()] = op->value;
+    return IRVisitor::Visit_(op);
+  }
+
+  void Visit_(const AttrStmt *op) final {
+    if (!InThreadEnv() && (op->attr_key == attr::thread_extent ||
+                           op->attr_key == attr::pipeline_exec_scope)) {
+      EnterThreadEnv();
+      IRVisitor::Visit_(op);
+      ExitThreadEnv();
+    } else {
+      IRVisitor::Visit_(op);
+    }
+  }
+
+  void Visit_(const ProducerConsumer *op) final {
+    EnterProducerConsumer(op);
+    IRVisitor::Visit_(op);
+    ExitProducerConsumer();
+  }
+
+  void Visit_(const Load *op) final {
+    HandleLoadStoreToVariable(op->buffer_var);
+    return IRVisitor::Visit_(op);
+  }
+
+  void Visit_(const Store *op) final {
+    HandleLoadStoreToVariable(op->buffer_var);
+    return IRVisitor::Visit_(op);
+  }
+  //@}
+
+  /// Check if the value of a Variable comes from function argument.
+  bool IsFromFunctionArgs(const Variable *var) const {
+    const Variable *V = var;
+    while (1) {
+      CHECK(V) << "Invalid Variable\n";
+
+      // Variable is from function args. Return true.
+      if (V == func_->args[0].node_.get()) return true;
+
+      // The value is expected to come from a tvm_struct_get Call.
+      // Get the first argument of tvm_struct_get, and continue.
+      const auto &iter = defs_.find(V);
+      if (iter == defs_.end()) return false;
+      const Call *C = iter->second.as<const Call>();
+      if (!C || C->name != intrinsic::tvm_struct_get) return false;
+      V = C->args[0].as<Variable>();
+    }
+    return false;
+  }
+
+  /// Handle memory access to a Variable
+  void HandleLoadStoreToVariable(const VarExpr &var) {
+    // We skip the access within thread env.
+    if (InThreadEnv()) return;
+
+    // We only check access within a producer/consumer.
+    // Because for load/store out side of producer/consumer,
+    // they don't have to be in thread env to stay legal (e.g. Load of args).
+    if (!InProducerConsumer()) return;
+
+    // We only handle the variable from function argument.
+    // If it does not come from args, then it could be allocated internally,
+    // it may possibly be in host or device address space.
+    // We do not handle this case, and skip it conservatively.
+    if (!IsFromFunctionArgs(var.get())) return;
+
+    // The verification fails in this case.
+    SetFailure();
+  }
+
+  /// Status getter/setter
+  //@{
+  inline bool InThreadEnv() const { return in_thread_env_; }
```

Comment:
no need to mark member function as inline if they are defined in the scope, they are already inlined by default